server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt

```diff
@@ -186,23 +209,30 @@ object Chapter {
                     ChapterTable.select { ChapterTable.manga eq mangaId }
                         .orderBy(ChapterTable.url to ASC).toList()
                 }
+
             val chapterUrls = chapterList.map { it.url }.toSet()
+            val chaptersToDelete = mutableListOf<ResultRow>()
 
             dbChapterList.forEachIndexed { index, dbChapter ->
-                if (
-                    !chapterUrls.contains(dbChapter[ChapterTable.url]) || // is orphaned
-                    (index < dbChapterList.lastIndex && dbChapter[ChapterTable.url] == dbChapterList[index + 1][ChapterTable.url]) // is duplicate
-                ) {
-                    transaction {
-                        PageTable.deleteWhere { PageTable.chapter eq dbChapter[ChapterTable.id] }
-                        ChapterTable.deleteWhere { ChapterTable.id eq dbChapter[ChapterTable.id] }
-                    }
+                val isOrphaned = !chapterUrls.contains(dbChapter[ChapterTable.url])
+                val isDuplicate =
+                    index < dbChapterList.lastIndex && dbChapter[ChapterTable.url] == dbChapterList[index + 1][ChapterTable.url]
+
+                if (isOrphaned || isDuplicate) {
+                    chaptersToDelete.add(dbChapter)
                 }
             }
+
+            val chaptersIdsToDelete = chaptersToDelete.map { it[ChapterTable.id] }
+
+            transaction {
+                PageTable.deleteWhere { PageTable.chapter inList chaptersIdsToDelete }
+                ChapterTable.deleteWhere { ChapterTable.id inList chaptersIdsToDelete }
+            }
         }
 
         if (manga.inLibrary) {
-            downloadNewChapters(mangaId, numberOfCurrentChapters, newChapters)
+            downloadNewChapters(mangaId, chapterList.size, newChapters)
```

Comment:
the second argument has to be the previous number of chapters, otherwise, it won't be possible to detect new chapters